README.md
invenio-arsharma
================

development branch for: 
favorite-collections &amp; CDS query analysis + visualization

**This repo contains 2 sub-repos:**
     1. invenio_favorite_collections
     2. invenio_search_viz

     1. contains lib folder for invenio's websearch module, 
        which was modified for adding favorites feature

     2. contains a separate structure containing directories: 
      >> data : this one is missing query_dict.data 
      	      	and user_query_tuple.data (they're over 150Mb's)
      >> public
      >> src
      >> docs

additionally a docs folder is included to support documentation and development flow

docs/changeLog.txt
1. Webstyle:
   a) webstyle_templates.py:
 			- def tmpl_pageheader():
			        - new jquery, jquery-ui link

2. Websearch:
   a) websearch_favorite_collection.py
   b) websearch_templates.py :
     			- def tmpl_favorite_collection()
			- def tmpl_search_query_viz()
     			- def tmpl_narrowsearch():
				- #HTML for favorites hovercard: in sons & grandsons
   c) websearch_webcoll.py:
			- def perform_display_collection():
   			        - fav_data
			        - fav_body
			        - c_body (,,,fav_body,,)

   d) websearch_webinterface.py:
			# Resulting in new urls /hovercard & /search_q_viz
			- class WebInterfaceSearchInterfacePages():
			  	- _exports = [,,'search_q_viz', 'hovercard',,,]
				- def search_q_viz()
				- def hovercard()
								
3. /var/www/static/:
	- data.txt : to test /static readability
	- data.json: sample data to pick up into visualizations

4. invenio DB:
   	- fav_collec : New table to keep user favorite collection preferences

===============================================
> webinterface_webcoll.py

    # import Db_logic_function from invenio.websearch.websearch_favorites                                                                          
    # Db_logic_function(uID) (uses dbquery module):                                                                                                
    # returns the tuple for a uID (if User is present) displaying collection name and id in a tuple                                                
    # adds a collID with a uID is user authenticated and requires tuple to be added                                                                
    # deletes a collID with a uID if deletion requested                                                                                            

    # then add here the fav_body()                                                                                                                 
    # where the fav_body calls the tmpl_favorites() sending the tuple and where the HTML for favorites is formed.                                  

    # in c_body add to the passed parameters, the data object for HTML returned from tmpl_favorites.                                               
    # this c_body now contains the HTML for favorites included (if uID is authentic)                                                               
    # this c_body now goes back to web_interface script and becomes part of URL which opens the main webpage.                                      

    # meanwhile,                                                                                                                                   
    # tmpl_narrowsearch prepares the HTML for focuson, so similarly, pass a fav_data_object which adds some <div> for hovercards for each collecti\
on                                                                                                                                                 
    # when deletion required, (a cross displayed in front of favorites), an AJAX request passed to webinterface file, (in req) and collID and uID \
is passed. this gets passed to websearch_favorites which calls deleted method and returns the new favorites tuple. this return tuple, when bibsche\
d webcoll is run, updates the display  

========================================================


invenio_search_viz/src/websearch_visualizer_json_parser.py
## This file is part of Invenio.
## Copyright (C) 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013 CERN.
##
## Invenio is free software; you can redistribute it and/or
## modify it under the terms of the GNU General Public License as
## published by the Free Software Foundation; either version 2 of the
## License, or (at your option) any later version.
##
## Invenio is distributed in the hope that it will be useful, but
## WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
## General Public License for more details.
##
## You should have received a copy of the GNU General Public License
## along with Invenio; if not, write to the Free Software Foundation, Inc.,
## 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.

""" test cmd script to show frequency of fields from a particular week's json file"""

import json, operator, pprint

weeks = range(22,36)
print "Weeks for which data is available:\n",weeks,"\n"
week = int(raw_input('Enter the week #, for loading json: '))

try:
    f = open('../data/json/count_'+str(week)+'.json','r')
    freq_data = json.load(f)
    print freq_data.keys()
    f.close()
except:
    if week not in weeks:
        quit("Wrong week specified!\nQuitting..")
    else:
        quit("No JSON object could be decoded.\nMaybe due to a previous UTF-8 dump error through json.dumps().\nQuitting..")

# and sort values
raw_input('\npress enter to proceed... ')

fields = ['c','a','p']
choice = raw_input("\n----\n'a' - author\n\'c' - collection\n'p' - global searched terms\n-----\nEnter your choice(letter): ")
print type(choice)
try:
    sorted_data = sorted(freq_data[choice].iteritems(), key=operator.itemgetter(1))
    pprint.pprint(sorted_data)
except:
    print '\nwrong choice!\n'
    quit()


invenio_search_viz/src/webstat_visualizer_csv_generator.py
## This file is part of Invenio.
## Copyright (C) 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013 CERN.
##
## Invenio is free software; you can redistribute it and/or
## modify it under the terms of the GNU General Public License as
## published by the Free Software Foundation; either version 2 of the
## License, or (at your option) any later version.
##
## Invenio is distributed in the hope that it will be useful, but
## WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
## General Public License for more details.
##
## You should have received a copy of the GNU General Public License
## along with Invenio; if not, write to the Free Software Foundation, Inc.,
## 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.

""" Search Query URL arguments processor """

#import dependencies
import pandas as pd
#from urlparse import parse_qs
import gzip, cPickle, re, json
#import numpy as np

try:
    from invenio.search_engine import create_basic_search_units
except:
    print '..Invenio not installed (properly). I quit!'
    quit()

f = gzip.open('../data/user_query_tuple.data','rb')
pd_tuple_list = pd.DataFrame.from_records(list(cPickle.load(f)))
f.close()

'''
f = gzip.open('../data/query_dict.data','rb')
pd_dict = pd.Series(cPickle.load(f))
f.close()
'''

#x = pd.DatetimeIndex(pd_tuple_list[2])
'''
x -> 
<class 'pandas.tseries.index.DatetimeIndex'>
[2012-06-01 00:00:01, ..., 2012-08-31 23:59:50]
Length: 1535896, Freq: None, Timezone: None

x.week -> 
array([22, 22, 22, ..., 35, 35, 35], dtype=int32)
'''

weeks = sorted(set(x.week))
print weeks
#[22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35]

index = 0
max_index = len(pd_tuple_list)
for week in weeks:
    f = open('../data/csv/w_'+ str(week)+'.csv', 'w')
    f.write('q_id,u_id,timestamp'+'\n')
    try:
        while(pd_tuple_list[2][index].week == week and index < max_index):
            f.write(str(pd_tuple_list[0][index]) + ',' 
                    + str(pd_tuple_list[1][index]) + ',' 
                    + pd_tuple_list[2][index].isoformat() + '\n')
            index+=1
        f.close()
        print week, index
    except:
        pass
'''
    for i in pd_tuple_list[2]:
    if i.week == week:
    tmp.append(i)
        else:
    break
'''


invenio_favorite_collections/lib/websearch_favorite_collection.py
## This file is part of Invenio.
## Copyright (C) 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013 CERN.
##
## Invenio is free software; you can redistribute it and/or
## modify it under the terms of the GNU General Public License as
## published by the Free Software Foundation; either version 2 of the
## License, or (at your option) any later version.
##
## Invenio is distributed in the hope that it will be useful, but
## WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
## General Public License for more details.
##
## You should have received a copy of the GNU General Public License
## along with Invenio; if not, write to the Free Software Foundation, Inc.,
## 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.

"""Favorite Collections feature"""

from invenio.dbquery import run_sql as res
from invenio.webuser import \
    getUid, page_not_authorized, \
    isGuestUser, collect_user_info, \
    isUserSubmitter, username_exists_p, \
    isUserAdmin, get_email, \
    get_session, get_user_preferences, get_user_info

from invenio.urlutils import redirect_to_url, make_canonical_urlargd, drop_default_urlargd
#import invenio.websearchadminlib as wa
#from websearch_webcoll import Collections
#from pexpect import spawn

#favorite_collection = res("SELECT * FROM fav_collec")
#user_info = collect_user_info(req)

def _get(req):
    uid = getUid(req)
    
def perform_add_new_favorite_from_collection(uID, collID, req):
    """Add collection to favorite(s).
    
    @param uID: User ID who wants to add the collection
    @param collID: collection ID for adding to favorites
    @param req: returned object from AJAX request
    """
    if isGuestUser(uID):
        return "login first!"
    else:
        try:
            favorites = res("INSERT INTO into fav_collec values(%d,%d)"%(uID,collID));
            return favorites
        except:
            return ''

def perform_delete_favorite_from_collection(uID, collID, req):
    """Delete a collection from favorite(s).
    @param uID: User ID who wants to add the collection
    @param collID: collection ID for adding to favorites
    @param req: returned object from AJAX request
    """
    if isGuestUser(uID):
        return "..login to access this feature!"
    else:
        try:
            favorites = res("DELETE FROM fav_collec WHERE id_user=%d AND id_collection=%d"%(uID, collID))
            return "Successfully deleted the requested collection from list of favorites!"
        except:
            return ''

def perform_calculate_favorites(uID):
    """Talks to DB for extracting favorites for a user
    
    @param uID: User ID of currently logged in user
    """
    if isGuestUser(uID):
        return "..login to access this feature!"
    else:
        try:
            id_favs = res("SELECT id_collection FROM fav_collec WHERE id_user=%d"%(uID))
            favorites = []
            for id_fav in id_favs:
                favorites.append(res("SELECT name FROM collection WHERE id=%d"%(id_fav[0]))[0][0])
            return favorites
        except:
            return ''


invenio_search_viz/src/webstat_visualizer_csv_processor.py
## This file is part of Invenio.
## Copyright (C) 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013 CERN.
##
## Invenio is free software; you can redistribute it and/or
## modify it under the terms of the GNU General Public License as
## published by the Free Software Foundation; either version 2 of the
## License, or (at your option) any later version.
##
## Invenio is distributed in the hope that it will be useful, but
## WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
## General Public License for more details.
##
## You should have received a copy of the GNU General Public License
## along with Invenio; if not, write to the Free Software Foundation, Inc.,
## 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.

import pandas as pd
from urlparse import parse_qs
import gzip, cPickle, json, operator
#import re
import numpy as np

try:
    f = gzip.open('../data/query_dict.data','rb')
    print "\n....loading query data. Hold on.."
    print "\n time before loading: %s"%(pd.datetime.today().time().isoformat())
    pd_dict = pd.Series(cPickle.load(f))
    f.close()
    print "\n time after loading: %s\n\n"%(pd.datetime.today().time().isoformat())
except:
    quit("\nError: invalid path specified for data file!\nQuitting..")

weeks = [ 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35] 
# right now declare no. of weeks manually because of sample data, 
# later on, store in some buffer or a file.

def perform_breakdown_query_units(q_parsed, f):
    """ Searches for combinations within parsed query
    For eg: after loading dictionary pd_dict, look up 
    1. parse_qs(pd_dict[3214343]) 
    2. parse_qs(pd_dict[10]) 
    3. parse_qs(pd_dict[6])
   
    Some of them might contain key 'f', some might contain keyword in key 'p'
    others might not contain anything, just plain 'p'
    
    @params q_parsed: parsed url arguments
    @params f: = 'keyword' -> field name to be passed to create_basic_search_units()
    """
    parsed_keys = q_parsed.keys()    
    p_tmp=[]; c_tmp=[]; f_tmp=[]; a_tmp=[]
    p_chk = False; f_chk= False #; flag = False
    # check for presence of url arguments 'p', 'c' and 'f'
    if 'p' in parsed_keys:
        p_tmp = q_parsed['p'] # query string
        p_chk = True
    if 'c' in parsed_keys:
        c_tmp = q_parsed['c'] # list of all collections present in url arguments
    if 'f' in parsed_keys:
        f_tmp = q_parsed['f'] # list of field names (if any)
        f_chk = True

    if p_chk:
        # check if a keyword is embedded in search query
        if len(p_tmp[0].split(':'))>1:
            # if ':' present, means a field is implicitly mentioned.
            p_breakdown = create_basic_search_units(None, p=p_tmp[0], f=f)

            for broken_args in p_breakdown:
                if broken_args[2] == 'author':
                    #flag = True # checked: author name implicitly included in query
                    if '*' in broken_args[1]:
                        # find position of arguments before/after astrix '*' 
                        # (sometimes included in author name)
                        pos_tmp = np.where(np.array(broken_args[1].split('*'))!='')[0]
                        #append the first argument in resultant array
                        a_tmp.append(broken_args[1].split('*')[pos_tmp[0]]) 
                    else:
                        # if in proper name format already, append already!
                        a_tmp.append(broken_args[1])
                else:
                    pass
        else:
            pass
    else:
        pass
    
    # next, check f_tmp for field names, if present, then check for 'author'
    if f_chk: 
        for i in f_tmp:
            if i=='author':
                a_tmp.append(p_tmp[0])
    # finally return everything, in 3 categorized lists
    return p_tmp, a_tmp, c_tmp

f = 'keyword' # declare argument for method: create_basic_search_units()

for week in weeks:
    collec_tmp = {}
    # read the respective csv files
    # Format: q_id, u_id, timestamp
    data = pd.read_csv('../data/csv/w_'+ str(week)+'.csv') #
    print "\n 1st 2 rows of data from csv of current week %d\n %s"%(week,data[:2]),"\n"
    print "Time: %s \n"%(pd.datetime.today().time().isoformat())
    #q_ids = data['q_id'].values

    #for q_id in q_ids:
    for elem in xrange(len(data)):
        q_id = data.get_value(elem, 'q_id')
        try:
            query_parsed =  parse_qs(pd_dict[q_id]) # dictionary with field as keys
            
            # categorize parameters, and return as lists
            p_tmp, a_tmp, c_tmp = perform_breakdown_query_units(query_parsed, f)
            
            # make list of tuples with (field, parameters)
            fields = [('c',c_tmp),('a',a_tmp),('p',p_tmp)]
            for field in fields:
                try:
                    # check if current field exists (as a dictionary)
                    collec_tmp[field[0]].keys()
                except:
                    # else make one
                    collec_tmp[field[0]]={}
                for element in field[1]:
                    if element in collec_tmp[field[0]].keys():
                        # parameter already present? ..increase count
                        collec_tmp[field[0]][element]+=1
                    else:
                        # else initialize count
                        collec_tmp[field[0]][element]=1
        except:
            pass
    for i in ['c','a','p']:
        print "\n..length of %s: %d\n"%(i,len(collec_tmp[i].keys()))
    try:
        f1 = open('../data/json/count_'+str(week)+'.json','w')
        f1.write(json.dumps(collec_tmp))
        f1.close()
    except:
        print "..couldnt dump json for week %d !!"%(week)
        print "..probably UTF-8 encoding error"
        pass

#print json.dumps(collec_tmp),"\n"
#dumpIt = unicode(str(collec_tmp), errors='ignore')
'''
# code to load dump 
f = open('count.json','r')
freq_data = json.loads(f)
f.close()
print freq_data.keys()

# and sort values
raw_input('press enter to print most searched author')
print sorted(freq_data['a'].iteritems(), key=operator.itemgetter(1)) 
'''

#next
# save in json with embedded keys (dict of dic) and 1st key as timestamp
